### DISCUSSION TITLE
how to test a modification of crazyflie's PID?

### DISCUSSION THREAD

**Discussion by loraine92** [bitcraze/discussions#1004] [Q&A]
Could you please help me, I modified the PID of crazyflie in the virtual machine in the crazyflie-firmware folder, I want to know if when I run a script to test the PID in visual studio in the virtual machine, the script is executed with the modifications I made to the PID in the crazyflie-firmware folder, or should I also update the firmware on the drone?

**Comment by knmcguire** [comment]
Hi! so if you have a python script which changes the [PID parameters before flight](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#posctlpid) that is not persistently saved, than once the crazyflie takes off it will fly with those modifications. Once restart the crazyflie, it won't save those values so you'll need to indicate those values again with the same script.

Have you found the [tuning tab of the cfclient yet?](https://github.com/orgs/bitcraze/discussions/1004). There you can tune the crazyflie in air with slidebars and if you are happy you can 'persist' the parameter values. This means that the values are saved in a seperate memory on the crazyflie (eeprom) and these will be read and implemented at every restart.

These are some api documentation of how to set persistent parameters from your script: https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/crazyflie/param/

**Reply by loraine92** [reply]
I still have several questions, can you please help me.
The modifications were made not only in the PID parameters but also in the structural design of the PID, the modifications were based on the new PID design that was simulated in Matlab, which is shown in the figure and the modifications were implemented in the firmware files in the virtual machine, in the following location: /home/bitcraze/Desktop/projects/crazyflie-firmware/, I made the modifications using visualStudio, and modified the following files:
- pid.h
- pid.c
- position_controller_pid.c
- position_controller.h
- platform_defaults_bolt.h
- platform_defaults_cf2.h,
- platform_defaults_flapper.h
- platform_defaults_tag.h
If I run the autonomous_sequence_high_level.py script when the drone flies and makes the trajectory, does it use the controller that I designed and that is in the virtual machine in /home/bitcraze/Desktop/projects/crazyflie-firmware/? Should I update the firmware with the modifications I made to the PID on the drone from the cfclient in connect-> Bootloader? When is the autonomous_sequence_high_level.py script executed from visualStudio, which PID executes, the one saved on the drone or what is in the virtual machine in /home/bitcraze/Desktop/projects/crazyflie-firmware/?
What I am doing is for my master's thesis, I would like to know if you have done something similar and have an article
that you can recommend to me.

thank you so much
![PID_PARALELO](https://github.com/bitcraze/discussions/assets/80229306/123e4c49-fa36-46db-8b42-f5afd486de0a)

**Reply by knmcguire** [reply]
aha... yeah that complicates things a bit. You mean that you want to change the embedded PID controller on the crazyflie firmware. You will need to remake the crazyflie firmware from the source with those changes and upload it to the crazyflie. All the instructions for those are listed on the crazyflie-firmware repo docs: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/building-and-flashing/build/. These also consist of an explanation of how to reflash it from the terminal (which you have to do, not through the client).

Before changing things in the firmware, if you are new to embedded programming, perhaps step back and think if this is the right approach. You can also consider [sending the reference attitude ](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/python_api/#sending-control-setpoints-with-the-commander-framework)from your offboard controller to the crazyflie with attitude setpoints instead? Or at least as a first step. It would require you to rewrite the controller in python first but at least that is an easier step than to implement it embedded first (is a bit of the cold turkey approach).

We have quite some master theses [listed here](https://www.bitcraze.io/portals/research/). And you can perhaps also check out this paper: Socas, Rafael, et al. "A Control Engineering Framework for Quadrotors: An Application for the Crazyflie 2. X." ICINCO. 2021.

This is also a simple tutorial to explain how to get started with embedded development: https://www.bitcraze.io/documentation/tutorials/getting-started-with-development/.
